pkg/kubelet/kubelet_node_status.go

```diff
@@ -39,67 +39,117 @@ import (
 	"k8s.io/kubernetes/pkg/volume/util/volumehelper"
 )
 
-// registerWithApiserver registers the node with the cluster master. It is safe
+// registerWithApiServer registers the node with the cluster master. It is safe
 // to call multiple times, but not concurrently (kl.registrationCompleted is
 // not locked).
-func (kl *Kubelet) registerWithApiserver() {
+func (kl *Kubelet) registerWithApiServer() {
 	if kl.registrationCompleted {
 		return
 	}
 	step := 100 * time.Millisecond
+
 	for {
-		time.Sleep(step)
+		// wake up and calculate the next backoff
+		timer := time.NewTimer(step)
+		<-timer.C
 		step = step * 2
 		if step >= 7*time.Second {
 			step = 7 * time.Second
 		}
 
-		node, err := kl.initialNodeStatus()
+		node, err := kl.initialNode()
 		if err != nil {
 			glog.Errorf("Unable to construct api.Node object for kubelet: %v", err)
 			continue
 		}
 
-		glog.V(2).Infof("Attempting to register node %s", node.Name)
-		if _, err := kl.kubeClient.Core().Nodes().Create(node); err != nil {
-			if !apierrors.IsAlreadyExists(err) {
-				glog.V(2).Infof("Unable to register %s with the apiserver: %v", node.Name, err)
-				continue
-			}
-			currentNode, err := kl.kubeClient.Core().Nodes().Get(kl.nodeName)
-			if err != nil {
-				glog.Errorf("error getting node %q: %v", kl.nodeName, err)
-				continue
-			}
-			if currentNode == nil {
-				glog.Errorf("no node instance returned for %q", kl.nodeName)
-				continue
-			}
-			if currentNode.Spec.ExternalID == node.Spec.ExternalID {
-				glog.Infof("Node %s was previously registered", node.Name)
-				kl.registrationCompleted = true
-				return
-			}
-			glog.Errorf(
-				"Previously %q had externalID %q; now it is %q; will delete and recreate.",
-				kl.nodeName, node.Spec.ExternalID, currentNode.Spec.ExternalID,
-			)
-			if err := kl.kubeClient.Core().Nodes().Delete(node.Name, nil); err != nil {
-				glog.Errorf("Unable to delete old node: %v", err)
+		glog.Infof("Attempting to register node %s", node.Name)
+		registered := kl.registerWithApiServerAttempt(node)
+		if registered {
+			glog.Infof("Successfully registered node %s", node.Name)
+			kl.registrationCompleted = true
+			return
+		}
+	}
+}
+
+// registerWithApiServerAttempt makes an attempt to register the given node
+// with the API server, returning a boolean indicating whether the attempt was
+// successful.  If a node with the same name already exists, it reconciles the
+// value of the annotation for controller-managed attach-detach of attachable
+// persistent volumes for the node.  If a node of the same name exists but has
+// a different externalID value, it attempts to delete that node so that a
+// later attempt can recreate it.
+func (kl *Kubelet) registerWithApiServerAttempt(node *api.Node) bool {
+	_, err := kl.kubeClient.Core().Nodes().Create(node)
+	if err == nil {
+		return true
+	}
+
+	if !apierrors.IsAlreadyExists(err) {
+		glog.Errorf("Unable to register node %q with API server: %v", kl.nodeName, err)
```

Comment:
We're supposed to use `utilruntime.HandleError` instead of `glog.Error`, but looking at `pkg/kubelet`, I see that the latter happens much more than the former.